### PR TITLE
Issue 281: Schema Merge

### LINKED ISSUE #281: Appending data should not merge schemas at rename
## What went wrong?

The expected behavior is that:

- if I create a table with 5 columns, and then I try to append to it a dataset with 6 columns, it should fails. Delta fails, but Qbeast doesn't and merge the schema.
- if I create a table with 5 columns, and then we append a dataset that contains only part of them (let's say the first 4), the operation should succeed, and the last column is full of null. In this, both Delta and Qbeast behave correctly.
- if I create a a table with 5 columns, and then I try to append a dataset with 5 columns, but one of them it is renamed, it should fail. Here Qbeast fails to raise an exception and instead, merges the schema.

## How to reproduce?

Different steps about how to reproduce the problem.

### 1. Code that triggered the bug, or steps to reproduce:

```scala

val df = Seq((2073090619, 773622413, "PRT76gRvYP", "{8RxqrEd9TY}")).toDF("a", "b", "field_with_forbidden_characters_", "another_field_with_forbidden_characters")
df.write.format("qbeast").option("columnsToIndex", "a,b").save("/tmp/dir")

val renamed_df = df.withColumn("rename_field", col("another_field_with_forbidden_characters")
renamed_df.write.format("qbeast").mode("append").save("/tmp/dir")


```

The resulting schema is

`| a         | b         | field_with_forbidden_characters_ | another_field_with_forbidden_characters | renamed_field |
`

### 2. Branch and commit id:

`1.0.0-main` 

### 3. Spark version:
On the spark shell run `spark.version`.

`3.5.0`

### 4. Hadoop version:
On the spark shell run `org.apache.hadoop.util.VersionInfo.getVersion()`.

### 5. How are you running Spark?
Are you running Spark inside a container? Are you launching the app on a remote K8s cluster? Or are you just running the tests in a local computer?

Locally

### 6. Stack trace:
Trace of the log/error messages.

### PR DESCRIPTION
**Pull request by osopardo1**
## Description

Fixes #281 . Schema changes in the data where merged by default when appending it to an existing table.

Two main problems are solved in the PR:

1. The `QbeastMetadataOperation` is no longer setting up `canMergeSchema` and `canOverwriteSchema` parameters to `true` by default. Now they are initialized according to the properties specified in `DeltaOptions` / `QbeastOptions`.
3. When Creating a Table, the behavior of Delta is to do a **first commit in the path with only the Metadata** (schema, table id...). In the Qbeast case, no initial schema was committed. This is solved by introducing a **first write with an empty data frame and a Convert To Qbeast operation**.

## Type of change

Bug fix. No breaking changes. Change only qbeast-spark package.

## Checklist:

Here is the list of things you should do before submitting this pull request:

- [x] New feature / bug fix has been committed following the [Contribution guide](https://github.com/Qbeast-io/qbeast-spark/blob/main/CONTRIBUTING.md).
- [x] Add comments to the code (make it easier for the community!).
- [x] Add tests.
- [x] Your branch is updated to the main branch (dependent changes have been merged).

## How Has This Been Tested? (Optional)

Added test to `QbeastSchemaTest`.

**Test Configuration**:
* Spark Version: 3.5.0
* Hadoop Version: 3.3.3
* Cluster or local? Local

### CODE REVIEW

File: src/test/scala/io/qbeast/spark/utils/QbeastSchemaTest.scala [hunk 1]
```diff
@@ -131,4 +131,90 @@ class QbeastSchemaTest extends QbeastIntegrationTestSpec {
...
+
+      import spark.implicits._
+
+      spark.sql(
+        "CREATE TABLE student (id INT) USING qbeast " +
+          "OPTIONS ('columnsToIndex'='id')")
+
+      val dfExtraCol = Seq((1, "John"), (2, "Doe")).toDF("id", "name")
+
+      // EXRA COLUMN
```
Comment (Jiaweihu08) [inline src/test/scala/io/qbeast/spark/utils/QbeastSchemaTest.scala:154]: EXTRA

File: src/test/scala/io/qbeast/spark/utils/QbeastSchemaTest.scala [hunk 2]
```diff
@@ -131,4 +131,90 @@ class QbeastSchemaTest extends QbeastIntegrationTestSpec {
...
+      val path = s"$tmpDir/student"
+      df.write
+        .format("qbeast")
+        .mode("overwrite")
+        .option("columnsToIndex", "id")
+        .save(path)
+
+      val dfExtraCol = Seq((1, "John"), (2, "Doe")).toDF("id", "name")
+
+      // EXRA COLUMN
```
Comment (Jiaweihu08) [inline src/test/scala/io/qbeast/spark/utils/QbeastSchemaTest.scala:188]: EXTRA

**Review by Jiaweihu08 — APPROVED** [review]

**Comment by osopardo1** [comment]
Fixing the comments and adding coverage to `QbeastOptions` case class.

File: src/main/scala/io/qbeast/spark/internal/QbeastOptions.scala [hunk 3]
```diff
@@ -38,14 +40,44 @@ import org.apache.spark.sql.SparkSession
...
+      options += DeltaOptions.TXN_VERSION -> txnVersion
+    }
+    if (userMetadata.nonEmpty) {
+      options += DeltaOptions.USER_METADATA_OPTION -> userMetadata.get
+    }
+    if (mergeSchema.nonEmpty) {
+      options += DeltaOptions.MERGE_SCHEMA_OPTION -> mergeSchema.get
+    }
+    if (overwriteSchema.nonEmpty) {
+      options += DeltaOptions.OVERWRITE_SCHEMA_OPTION -> overwriteSchema.get
```
Comment (fpj) [inline src/main/scala/io/qbeast/spark/internal/QbeastOptions.scala:71]: It sounds like codecov is not happy about this line 71. Where do we exercise this `toMap` method in the tests?

File: src/main/scala/io/qbeast/spark/internal/QbeastOptions.scala [hunk 4]
```diff
@@ -135,14 +175,24 @@ object QbeastOptions {
     val txnAppId = getTxnAppId(options)
     val txnVersion = getTxnVersion(options)
     val userMetadata = getUserMetadata(options)
-    QbeastOptions(columnsToIndex, desiredCubeSize, stats, txnAppId, txnVersion, userMetadata)
+    val mergeSchema = getMergeSchema(options)
+    val overwriteSchema = getOverwriteSchema(options)
+    QbeastOptions(
```
Comment (fpj) [inline src/main/scala/io/qbeast/spark/internal/QbeastOptions.scala:180]: Perhaps not for this PR, but I'm wondering whether we should implement QbeastOptions following the builder pattern. It makes the creation of the object easier to maintain and visualize.

**Review by fpj — CHANGES_REQUESTED** [review]
Looks good, I only have a few questions.

File: src/main/scala/io/qbeast/spark/delta/DeltaMetadataWriter.scala [hunk 5]
```diff
@@ -66,6 +66,11 @@ private[delta] case class DeltaMetadataWriter(
 
   private def isOverwriteOperation: Boolean = mode == SaveMode.Overwrite
 
+  override protected val canMergeSchema: Boolean = options.canMergeSchema
+
+  override protected val canOverwriteSchema: Boolean =
+    options.canOverwriteSchema && isOverwriteOperation && options.replaceWhere.isEmpty
```
[the remark this answers is not in the document]
Comment (osopardo1) [reply src/main/scala/io/qbeast/spark/delta/DeltaMetadataWriter.scala:72]: This is part of Delta's code, but basically the option `canOverwriteSchema` is just an argument that allows to overwrite it, but unless the other requirements are met (which needs to be explicit), the enforcement of the schema is applied.

File: src/main/scala/io/qbeast/spark/internal/QbeastOptions.scala (hunk 3 above)
Comment (osopardo1) [reply src/main/scala/io/qbeast/spark/internal/QbeastOptions.scala:71]: Fixed on bd53be0 !

File: src/main/scala/io/qbeast/spark/internal/QbeastOptions.scala (hunk 4 above)
Comment (osopardo1) [reply src/main/scala/io/qbeast/spark/internal/QbeastOptions.scala:180]: Makes sense. Let me open an issue.

File: src/main/scala/io/qbeast/spark/delta/DeltaMetadataWriter.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (fpj) [reply src/main/scala/io/qbeast/spark/delta/DeltaMetadataWriter.scala:72]: Thank you, @osopardo1

**Review by fpj — APPROVED** [review]
+1, lgtm

**Pull request merged**